experiment/simulation/js/main.js

```diff
@@ -1 +1,577 @@
-//Your JavaScript goes in here
+'use strict';
+
+document.addEventListener('DOMContentLoaded', function() {
+
+	const restartButton = document.getElementById('restart');
+	const instrMsg = document.getElementById('procedure-message');
+
+	restartButton.addEventListener('click', function() { restart(); });
+
+	function finish(step)
+	{
+		if(!flag && step === enabled.length - 1)
+		{
+			flag = true;
+			logic(tableData);
+			generateTableHead(table, Object.keys(tableData[0]));
+			generateTable(table, tableData);
+
+			document.getElementById("output4").innerHTML = "Outflow Volume, Q = " + String(outflow) + " cm<sup>3</sup>";
+			document.getElementById("apparatus").style.display = 'none';
+			document.getElementById("observations").style.width = '40%';
+			if(small)
+			{
+				document.getElementById("observations").style.width = '85%';
+			}
+		}
+	};
+
+	function randomNumber(min, max) {
+		return Number((Math.random() * (max - min + 1) + min).toFixed(2));
+	};
+
+	function randomInt(min, max) {
+		return Number(Math.floor(randomNumber(min, max)));
+	};
+
+	function logic(tableData)
+	{
+		const heads = [30, 50, 60, 70], times = [randomInt(80, 85), randomInt(50, 55), randomInt(45, 50), randomInt(35, 40)];
+		tableData.forEach(function(row, index) {
+			row['Trial No.'] = index;
+			row['Constant Head, h(cm)'] = heads[index];
+			row['Elapsed Time, t(sec)'] = times[index];
+			row['Permeability, K(cm/sec)'] = (outflow * length / (area * times[index] * heads[index])).toFixed(3);
+		});
+	};
+
+	function flow(obj, change, lim) {
+		if(obj.waterHeight >= lim)
+		{
+			return 1;
+		}
+
+		obj.waterHeight += change;
+		return 0;
+	};
+
+	class container {
+		constructor(height, width, x, y) {
+			this.height = height;
+			this.width = width;
+			this.pos = [x, y];
+			this.waterHeight = 0;
+			this.img = new Image();
+			this.img.src = './images/container.png';
+			this.img.onload = () => { ctx.drawImage(this.img, this.pos[0], this.pos[1], this.width, this.height); };
+		};
+
+		draw(ctx) {
+			ctx.fillStyle = "#1ca3ec";
+			ctx.beginPath();
+			ctx.rect(this.pos[0] + 5, this.pos[1] + this.height - 5, this.width - 10, -this.waterHeight);
+			ctx.closePath();
+			ctx.fill();
+
+			ctx.drawImage(this.img, this.pos[0], this.pos[1], this.width, this.height);
+		};
+	};
+
+	class soil {
+		constructor(height, width, x, y) {
+			this.height = height;
+			this.width = width;
+			this.pos = [x, y];
+			this.waterHeight = 0;
+		};
+
+		draw(ctx) {
+			ctx.beginPath();
+			ctx.fillStyle = "#654321";
```

Comment:
Colors may be moved to an enum in a separate js files with mnemonic keys for the names and codes for values. Will be easier to understand and maintain.